src/tools/pkcs11-tool.c

```diff
@@ -1233,6 +1254,53 @@ int main(int argc, char * argv[])
 	if (action_count == 0)
 		util_print_usage_and_die(app_name, options, option_help, NULL);
 
+	if (opt_uri) {
+		/* Check that no interfering options were set */
+		if (opt_module || opt_slot_set || opt_slot_index_set || opt_slot_description ||
+				opt_pin || opt_object_label || opt_object_class_str || opt_object_id_len) {
+			fprintf(stderr, "Invalid combination of arguments when --uri used\n");
+			util_print_usage_and_die(app_name, options, option_help, NULL);
+		}
+		if (opt_uri->slot_id) {
+			opt_slot = (CK_SLOT_ID)strtoul(opt_uri->slot_id, NULL, 0);
+			opt_slot_set = 1;
+		}
+		opt_module = opt_uri->module_path;
+		util_get_pin(opt_uri->pin, &opt_pin);
+		util_get_pin(opt_uri->pin_source, &opt_pin);
+		if (opt_pin)
+			opt_login = 1;
+		opt_object_label = opt_uri->object;
+		if (opt_uri->type) {
+			if (strcmp(opt_uri->type, "cert") == 0) {
+				opt_object_class = CKO_CERTIFICATE;
+				opt_object_class_str = "cert";
+			} else if (strcmp(opt_uri->type, "private") == 0) {
+				opt_object_class = CKO_PRIVATE_KEY;
+				opt_object_class_str = "privkey";
+			} else if (strcmp(opt_uri->type, "secret-key") == 0) {
+				opt_object_class = CKO_SECRET_KEY;
+				opt_object_class_str = "secrkey";
+			} else if (strcmp(opt_uri->type, "public") == 0) {
+				opt_object_class = CKO_PUBLIC_KEY;
+				opt_object_class_str = "pubkey,";
+			} else if (strcmp(opt_uri->type, "data") == 0) {
+				opt_object_class = CKO_DATA;
+				opt_object_class_str = "data";
+			} else {
+				fprintf(stderr, "Unsupported object type \"%s\" specified in PKCS#11 URI\n", opt_uri->type);
+				util_print_usage_and_die(app_name, options, option_help, NULL);
+			}
+		}
+		if (opt_uri->id) {
+			opt_object_id_len = opt_uri->id_len;
+			memcpy(opt_object_id, opt_uri->id, opt_object_id_len);
+		}
+	}
```

Comment:
Doing this allows us to have the change in the rest of the code quite small, but I thought that this could be done the other way round -- when uri is not provided, it can be constructed from the old options. By having the uri structure as the uniform exchange format in the pkcs11-tool accepted by all the operations, we could get rid of the number of different parameters to each of the function and just operate on the uri.

Indeed this would take more work and more changes to do that in all places so keeping this as a suggestion. If you believe it is a bad idea, feel free to discard ;)